nucliadb_node/nucliadb_node/indexer.py

```diff
@@ -0,0 +1,454 @@
+        cls, *, index_message: IndexMessage, nats_msg: Msg, mpu: MessageProgressUpdater
+    ):
+        instance = super().__new__(cls)
+        instance.index_message = index_message
+        instance.nats_msg = nats_msg
+        instance.mpu = mpu
+        return instance
+
+    @classmethod
+    def from_msg(cls, msg: Msg) -> "WorkUnit":
+        mpu = MessageProgressUpdater(msg, nats_consumer_settings.nats_ack_wait * 0.66)
+
+        pb = IndexMessage()
+        pb.ParseFromString(msg.data)
+
+        return cls.__new__(cls, index_message=pb, nats_msg=msg, mpu=mpu)
+
+    @property
+    def seqid(self) -> int:
+        return int(self.nats_msg.reply.split(".")[5])
+
+    @property
+    def shard_id(self) -> str:
+        return self.index_message.shard
+
+    @cached_property
+    def _sortable_id(self) -> tuple[int, int]:
+        # Priority based on message source and smaller seqid
+        source_priority = self.priorities[self.index_message.source]
+        return (source_priority, self.seqid)
+
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, WorkUnit):
+            return NotImplemented  # pragma: no cover
+        return self._sortable_id.__eq__(other._sortable_id)
+
+    def __lt__(self, other):
+        if not isinstance(other, WorkUnit):
+            return NotImplemented  # pragma: no cover
+        return self._sortable_id.__lt__(other._sortable_id)
+
+
+class ConcurrentShardIndexer:
+    def __init__(self, writer: Writer, node_id: str):
+        self.writer = writer
+        self.node_id = node_id
+        self.indexers: dict[str, tuple["PriorityIndexer", asyncio.Task]] = {}
+
+    async def initialize(self):
+        self.storage = await get_storage(service_name=SERVICE_NAME)
+
+    async def finalize(self):
+        tasks = []
+        reverse_indexers = {}
+        for shard_id, (_, task) in self.indexers.items():
+            tasks.append(task)
+            reverse_indexers[task] = shard_id
+
+        if tasks:
+            # REVIEW timeout. Is 10 a good timeout? Should we implement some
+            # indexer finalize policy? Nak everything and close, for example?
+            done, pending = await asyncio.wait(tasks, timeout=10)
+            if pending:
+                logger.error(
+                    f"{len(pending)} pending tasks are still running after concurrent shard indexer finalizes!"
+                )
+
+            for task in done:
+                error = task.exception()
+                if error is not None:
+                    shard_id = reverse_indexers[task]
+                    logger.exception(
+                        f"Indexer task for shard {shard_id} raised an error while finalizing it",
+                        exc_info=error,
+                    )
+
+        self.indexers.clear()
+
+    def index_message_soon(self, msg: Msg):
+        work = WorkUnit.from_msg(msg)
+
+        logger.info(
+            f"Index message for shard {work.shard_id} is being enqueued",
```

Comment:
`work.shard_id` is already in extra

Reply:
The advantage of not having variables in log messages is that you can group by those values.